Mathlib/Tactic/Linter/DirectoryDependency.lean

```diff
@@ -0,0 +1,487 @@
+  f n <|> match n with
+    | anonymous => none
+    | str n' _ => n'.findPrefix f
+    | num n' _ => n'.findPrefix f
+
+/-- Make a `NameSet` containing all prefixes of `n`. -/
+def Lean.Name.prefixes (n : Name) : NameSet :=
+  NameSet.insert (n := n) <| match n with
+    | anonymous => ∅
+    | str n' _ => n'.prefixes
+    | num n' _ => n'.prefixes
+
+/-- Collect all prefixes of names in `ns` into a single `NameSet`. -/
+def Lean.Name.collectPrefixes (ns : Array Name) : NameSet :=
+  ns.foldl (fun ns n => ns.union n.prefixes) ∅
+
+/-- Find a name in `ns` that starts with prefix `p`. -/
+def Lean.Name.prefixToName (p : Name) (ns : Array Name) : Option Name :=
+  ns.find? p.isPrefixOf
+
+/-- Find the dependency chain, starting at a module that imports `imported`, and ends with the
+current module. -/
+def Lean.Environment.importPath (env : Environment) (imported : Name) : Array Name := Id.run do
+  let mut result := #[]
+  let modData := env.header.moduleData
+  let modNames := env.header.moduleNames
+  if let some idx := env.getModuleIdx? imported then
+    let mut target := imported
+    for i in [idx.toNat + 1 : modData.size] do
+      if modData[i]!.imports.any (·.module == target) then
+        target := modNames[i]!
+        result := result.push modNames[i]!
+  return result
+
+namespace Mathlib.Linter
+
+open Lean Elab Command
+
+/--
+The `directoryDependency` linter detects detects imports between directories that are supposed to be
+independent. If this is the case, it emits a warning.
+-/
+register_option linter.directoryDependency : Bool := {
+  defValue := true
+  descr := "enable the directoryDependency linter"
+}
+
+namespace DirectoryDependency
+
+/-- `NamePrefixRel` is a datatype for storing relations between name prefixes.
+
+That is, `R : NamePrefixRel` is supposed to answer given names `(n₁, n₂)` whether there are any
+prefixes `n₁'` of `n₁` and `n₂'` of `n₂` such that `n₁' R n₂'`.
+
+The current implementation is a `NameMap` of `NameSet`s, testing each prefix of `n₁` and `n₂` in
+turn. This can probably be optimized.
+-/
+def NamePrefixRel := NameMap NameSet
+
+namespace NamePrefixRel
+
+instance : EmptyCollection NamePrefixRel := inferInstanceAs (EmptyCollection (NameMap _))
+
+/-- Make all names with prefix `n₁` related to names with prefix `n₂`. -/
+def insert (r : NamePrefixRel) (n₁ n₂ : Name) : NamePrefixRel :=
+  match r.find? n₁ with
+    | some ns => NameMap.insert r n₁ (ns.insert n₂)
+    | none => NameMap.insert r n₁ (.insert ∅ n₂)
+
+/-- Convert an array of prefix pairs to a `NamePrefixRel`. -/
+def ofArray (xs : Array (Name × Name)) : NamePrefixRel :=
+  xs.foldl (init := ∅)
+    fun r (n₁, n₂) => r.insert n₁ n₂
+
+/-- Get a prefix of `n₁` that is related to a prefix of `n₂`. -/
+def find (r : NamePrefixRel) (n₁ n₂ : Name) : Option (Name × Name) :=
+  n₁.findPrefix fun n₁' => do
+    let ns ← r.find? n₁'
+    n₂.findPrefix fun n₂' =>
+      if ns.contains n₂'
+      then (n₁', n₂')
+      else none
```

Comment:
I'm not a huge fan of this style for `if` as an expression, but apparently it is a common thing. So:
```suggestion
      if ns.contains n₂' then
        (n₁', n₂')
      else
        none
```